clang/runtime/dpct-rt/include/dpct/graph.hpp

```diff
@@ -0,0 +1,22 @@
+//==---- graph.hpp ----------------------------*- C++ -*----------------==//
+//
+// Copyright (C) Intel Corporation
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/ext/oneapi/experimental/graph.hpp>
+#include <sycl/sycl.hpp>
+
+namespace dpct {
+namespace experimental {
+
+typedef std::shared_ptr<sycl::ext::oneapi::experimental::command_graph<
+    sycl::ext::oneapi::experimental::graph_state::modifiable>>
+    command_graph_t;
```

Comment:
```suggestion
using graph_ptr = sycl::ext::oneapi::experimental::command_graph<
    sycl::ext::oneapi::experimental::graph_state::modifiable>*;

typedef std::shared_ptr<graph_ptr> command_graph_t;
```
I think underlying object needs to be a `graph_ptr` similar to `queue_ptr` or `event_ptr`. That could be initialized as follows (aka `cudaGraphCreate`):
```
*(graph) = new sycl::ext::oneapi::experimental::command_graph<sycl::ext::oneapi::experimental::graph_state::modifiable>(q);
```
Whereas q could be a global `sycl::queue`, it's a required parameter. We also have an overload for `sycl::context` and `sycl::device`.

Reply:
Agree that command_graph_t should be a pointer. 
```typedef shared_ptr<command_graph<graph_state::modifiable>> command_graph_t```
Declares this as a shared pointer. Is there a need for a raw pointer to a shared_ptr as suggested in your example?

Reply:
Not really. My suggestion was based on the fact that `queue_ptr` and `event_ptr` are raw pointer, correct? And that the mappings should be aligned.
I like the idea to reflect this in the name as suggested below https://github.com/oneapi-src/SYCLomatic/pull/2026/files#r1623938989
Is there any problem of using only a raw pointer here instead of the shared_ptr that I overlook?
```
typedef sycl::ext::oneapi::experimental::command_graph<
    sycl::ext::oneapi::experimental::graph_state::modifiable>* command_graph_ptr;
typedef sycl::ext::oneapi::experimental::command_graph<
    sycl::ext::oneapi::experimental::graph_state::executable>* command_graph_exec_ptr;
```

Reply:
A uniform handling of queues, events and graphs sounds good in general.

But by using a shared pointer we will not have to worry about releaseing the memory, in general. We can reset the `command_graph_ptr` instance, which in my opinion is a cleaner way to handled pointers. Thoughts?

Reply:
Good point. But I don't think lifetime is less of a concern here. 
Similar to` cudaStreamCreate` and `cudaStreamDestroy` pairs a well defined CUDA App will always follow a `cudaGraphDestroy` to a `cudaGraphCreate`. Same for `cudaGraphExecDestroy` ...

@EwanC what are you thoughts here?

Reply:
Consistency would generally be my preference, so if `queue_ptr` and `event_ptr` are raw pointers then aim for that. If shared pointers are preferable, which I don't object to, then could all of `command_graph_ptr`, `queue_ptr`, and `event_ptr` be a shared ptr? If so then adding `command_graph_ptr` as a raw pointer now, then updating all the ptr typedefs to be a shared pointer would be a path forward.